File: TopCoder/python/misc/MergeTwoSortedList.py

class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution:
    def mergeTwoLists(self, l1: ListNode, l2: ListNode) -> ListNode:
        head = ListNode(-1)
        temphead = head

        while l1 is not None and l2 is not None:
            if l1.val == l2.val:
                temphead.next = l1
                l1 = l1.next
                temphead.next.next = l2
                l2 = l2.next
                temphead = temphead.next.next
            elif l1.val < l2.val:
                temphead.next = l1
                l1 = l1.next
                temphead = temphead.next
            else:
                temphead.next = l2
                l2 = l2.next
                temphead = temphead.next

        while l1 is not None:
            temphead.next = l1
            l1 = l1.next
            temphead = temphead.next

        while l2 is not None:
            temphead.next = l2
            l2 = l2.next
            temphead = temphead.next

        return head.next

File: TopCoder/src/main/java/com/leetcode/linkedlist/MergeTwoSortedList.java
package com.leetcode.linkedlist;

  class ListNode {
      int val;
      ListNode next;
      ListNode(int x) { val = x; }
  }

public class MergeTwoSortedList {

    public ListNode mergeTwoLists(ListNode l1, ListNode l2) {

        ListNode head = new ListNode(-1);
        ListNode tempHead = head;

        while (l1 != null && l2 != null) {

            if (l1.val == l2.val) {
                tempHead.next = l1;
                l1 = l1.next;
                tempHead.next.next = l2;
                l2 = l2.next;
                tempHead = tempHead.next.next;
                tempHead.next = null;
            } else if (l1.val < l2.val) {
                tempHead.next = l1;
                l1 = l1.next;
                tempHead = tempHead.next;
            } else {
                tempHead.next = l2;
                l2 = l2.next;
                tempHead = tempHead.next;
            }
        }

        while (l1 != null) {
            tempHead.next = l1;
            l1 = l1.next;
        }

        while (l2 != null) {
            tempHead.next = l2;
            l2 = l2.next;
        }

        return head.next;
    }


    public static void main(String[] args) {
        ListNode l1 = new ListNode(1);
        l1.next = new ListNode(2);
        l1.next.next = new ListNode(4);

        ListNode l2 = new ListNode(1);
        l2.next = new ListNode(3);
        l2.next.next = new ListNode(4);
        ListNode result = new MergeTwoSortedList().mergeTwoLists(l1, l2);
        while (result != null) {
            System.out.println(result.val);
            result = result.next;
        }
    }



}
